src/profile-logic/process-profile.js

```diff
@@ -587,91 +591,147 @@ function _convertStackToCause(data: Object): Object {
 }
 
 /**
- * Process the markers by either converting stacks to causes, process the GC markers.
+ * Sometimes we don't want to extract a cause, but rather just the stack index
+ * from a gecko payload.
  */
-function _processMarkers(geckoMarkers: GeckoMarkerStruct): RawMarkerTable {
+function _convertPayloadStackToIndex(
+  data: MarkerPayload_Gecko
+): IndexIntoStackTable | null {
+  if (!data) {
+    return null;
+  }
+  if (data.stack && data.stack.samples.data.length > 0) {
+    const { samples } = data.stack;
+    return samples.data[0][samples.schema.stack];
+  }
+  return null;
+}
+
+/**
+ * Process the markers.
+ *  Convert stacks to causes.
+ *  Process GC markers.
+ *  Extract JS allocations into the JsAllocationsTable.
+ */
+function _processMarkers(
+  geckoMarkers: GeckoMarkerStruct
+): {|
+  markers: RawMarkerTable,
+  jsAllocations: JsAllocationsTable | null,
+|} {
+  const markers = getEmptyRawMarkerTable();
+  const jsAllocations = getEmptyJsAllocationsTable();
+  for (let markerIndex = 0; markerIndex < geckoMarkers.length; markerIndex++) {
+    const geckoPayload: MarkerPayload_Gecko = geckoMarkers.data[markerIndex];
+
+    if (geckoPayload && geckoPayload.type === 'JS allocation') {
+      // Build up a separate table for the JS allocation data, and do not
+      // include it in the marker information.
+      jsAllocations.time.push(geckoPayload.startTime);
+      jsAllocations.className.push(geckoPayload.className);
+      jsAllocations.typeName.push(geckoPayload.typeName);
+      jsAllocations.coarseType.push(geckoPayload.coarseType);
+      jsAllocations.duration.push(geckoPayload.size);
+      jsAllocations.inNursery.push(geckoPayload.inNursery);
+      jsAllocations.stack.push(_convertPayloadStackToIndex(geckoPayload));
+      jsAllocations.length++;
+
+      // Do not process the marker and add it to the marker list.
+      continue;
+    }
```

Comment:
So, I gave more thought during the night, and now I think that because this is derived data this may rather be computed in a selector.
The back story is that doing the computation at processing time is a mistake I did for the compare view: this is nice to get a prototype but this is a lot less flexible then. For the compare view I definitely plan to move everything to selectors.

For the allocations, the concern I have is that we might want also different tables and trees, like in the memory devtools panel. This table structure is perfectly adapted to what you wanted to display here, but is it adapted to the other possible visualizations we'd like to do in the future?

I think now is a good time to think about it, because once we ship it, we'll need an upgrade path to move back.

Reply:
I originally tried the selector approach, as this is derived data and failed at it. It increased the complexity a huge amount, especially with teaching the selectors how to do all the transform pipeline. I got lost in cross referencing things in the selector creation pipeline. Adding it to the thread simplified the approach, as then I only had to teach the transforms about this new property on the thread, as opposed to duplicating the transform pipeline to special case the allocation information. This abandoned approach combined the marker selectors, and thread selectors in confusing circular dependencies, which made the selector organization worse.

Thanks for challenging this approach, as it is a bit odd that it gets added to the `Thread` structure. As a follow-up I'm also going to re-evaluate how this information gets serialized. I'm considering doing this step on the actual serialization side on Gecko, so that this data gets added to the Thread directly, rather than serializing it out as a marker.

Luckily we can always revert this decision if it proves incorrect, as you say, with an upgrader.

Thanks for the review!